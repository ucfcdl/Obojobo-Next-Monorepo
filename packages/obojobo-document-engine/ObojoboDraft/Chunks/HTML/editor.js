import React from 'react'
import Common from 'Common'

import emptyNode from './empty-node.json'
import Icon from './icon'
import Node from './editor-component'
import Schema from './schema'
import Converter from './converter'

const HTML_NODE = 'ObojoboDraft.Chunks.HTML'

const plugins = {
	renderNode(props, editor, next) {
		switch (props.node.type) {
			case HTML_NODE:
				return <Node {...props} {...props.attributes} />
			default:
				return next()
		}
	},
	onKeyDown(event, editor, next) {
		const isHTML = editor.value.blocks.some(block => block.type === HTML_NODE)
		if (!isHTML) return next()

		switch (event.key) {
			// Insert a softbreak on enter
			case 'Enter':
				event.preventDefault()
				return editor.insertText('\n')

			case 'Tab':
				event.preventDefault()
				return editor.insertText('\t')

			default:
				return next()
		}
	},
	schema: Schema
}

Common.Store.registerEditorModel('ObojoboDraft.Chunks.HTML', {
	name: 'HTML',
	icon: Icon,
	isInsertable: true,
	insertJSON: emptyNode,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	plugins
})

const HTML = {
	name: HTML_NODE,
	components: {
		Node,
		Icon
	},
	helpers: {
		slateToObo: Converter.slateToObo,
		oboToSlate: Converter.oboToSlate
	},
	json: {
		emptyNode
	},
	plugins
}

export default HTML
