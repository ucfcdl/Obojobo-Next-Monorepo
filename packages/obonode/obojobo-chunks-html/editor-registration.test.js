import { Transforms } from 'slate'

import HTML from './editor-registration'
const HTML_NODE = 'ObojoboDraft.Chunks.HTML'

jest.mock('slate-react')

describe('HTML editor', () => {
	test('plugins.normalizeNode calls next if the node is not an ActionButton', () => {
		const next = jest.fn()
		HTML.plugins.normalizeNode([{}, []], {}, next)

		expect(next).toHaveBeenCalled()
	})

	test('plugins.normalizeNode calls next if all HTML children are text', () => {
		const button = {
			type: HTML_NODE,
			children: [{ text: '' }]
		}
		const next = jest.fn()

		HTML.plugins.normalizeNode([button, [0]], { children: [button] }, next)
		expect(next).toHaveBeenCalled()
	})

	test('plugins.normalizeNode calls Transforms on an invalid child', () => {
		jest.spyOn(Transforms, 'liftNodes').mockReturnValueOnce(true)

		const button = {
			type: HTML_NODE,
			children: [
				{
					type: 'mockElement',
					children: [{ text: '' }]
				}
			]
		}
		const editor = {
			isInline: () => false,
			children: [button]
		}
		const next = jest.fn()

		HTML.plugins.normalizeNode([button, [0]], editor, next)
		expect(Transforms.liftNodes).toHaveBeenCalled()
	})

	test('plugins.decorate exits when not relevent', () => {
		expect(HTML.plugins.decorate([{ text: 'mock text' }], {})).toMatchSnapshot()

		expect(HTML.plugins.decorate([{ children: [{ text: 'mock text' }] }], {})).toMatchSnapshot()
	})

	test('plugins.decorate renders a placeholder', () => {
		const editor = {
			children: [{ children: [{ text: '' }] }]
		}

		expect(HTML.plugins.decorate([{ children: [{ text: '' }] }, [0]], editor)).toMatchSnapshot()
	})

	test('plugins.onKeyDown deals with no special key', () => {
		const event = {
			key: 'k',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown([{}, [0]], {}, event)

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Enter]', () => {
		jest.spyOn(Transforms, 'insertText').mockReturnValueOnce(true)

		const event = {
			key: 'Enter',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown([{}, [0]], {}, event)
		expect(event.preventDefault).toHaveBeenCalled()
		expect(Transforms.insertText).toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Tab]', () => {
		jest.spyOn(Transforms, 'insertText').mockReturnValueOnce(true)
		const event = {
			key: 'Tab',
			preventDefault: jest.fn()
		}

		HTML.plugins.onKeyDown([{}, [0]], {}, event)

		expect(event.preventDefault).toHaveBeenCalled()
		expect(Transforms.insertText).toHaveBeenCalled()
	})

	test('plugins.renderNode renders HTML when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			element: {
				type: HTML_NODE,
				content: {}
			}
		}

		expect(HTML.plugins.renderNode(props)).toMatchSnapshot()
	})
})
