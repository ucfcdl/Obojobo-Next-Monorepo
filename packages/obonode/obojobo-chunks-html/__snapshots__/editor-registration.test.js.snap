// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML editor plugins.renderNode calls next 1`] = `undefined`;

exports[`HTML editor plugins.renderNode renders HTML when passed 1`] = `
<HTML
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.HTML",
    }
  }
/>
`;

exports[`HTML editor plugins.renderPlaceholder exits when not relevent 1`] = `undefined`;

exports[`HTML editor plugins.renderPlaceholder exits when not relevent 2`] = `undefined`;

exports[`HTML editor plugins.renderPlaceholder exits when not relevent 3`] = `undefined`;

exports[`HTML editor plugins.renderPlaceholder renders a placeholder 1`] = `
<span
  className="placeholder"
  contentEditable={false}
  data-placeholder="<!-- HTML code here -->"
/>
`;
